tariff/entsoe/api.go

```diff
@@ -0,0 +1,154 @@
+// Package entsoe implements a minimalized version of the European Network of Transmission System Operators for Electricity's
+// Transparency Platform API (https://transparency.entsoe.eu)
+package entsoe
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+	"sort"
+	"strconv"
+	"time"
+
+	"github.com/evcc-io/evcc/util/request"
+	"github.com/evcc-io/evcc/util/shortrfc3339"
+)
+
+// BaseURI is the root path that the API is accessed from.
+const BaseURI = "https://web-api.tp.entsoe.eu/api"
+
+// numericDateFormat is a time.Parse compliant formatting string for the numeric date format used by entsoe get requests.
+const numericDateFormat = "200601021504"
+
+var ErrInvalidData = errors.New("invalid data received")
+
+// DayAheadPricesRequest represents a helper struct for requesting 4.2.10 Day Ahead Prices [12.1.D]
+type DayAheadPricesRequest struct {
+	domain      DomainType
+	periodStart time.Time
+	periodEnd   time.Time
+}
+
+// ConstructDayAheadPricesRequest constructs a new DayAheadPricesRequest.
+// Domain and duration validity is not checked, that's on you.
+func ConstructDayAheadPricesRequest(domain DomainType, duration time.Duration) DayAheadPricesRequest {
+	// Round to the hour.
+	now := time.Now().Truncate(time.Hour)
+	return DayAheadPricesRequest{
+		domain:      domain,
+		periodStart: now,
+		periodEnd:   now.Add(duration),
+	}
+}
+
+// DoRequest requests the current Day Ahead Prices from ENTSOE.
+// The client is expected to provide a decorator to provide the security key.
+func (r *DayAheadPricesRequest) DoRequest(client *request.Helper) (PublicationMarketDocument, error) {
+	var res PublicationMarketDocument
+
+	// TODO This feels a bit silly, could we simplify this somehow?
```

Comment:
Use

    uri := fmt.Sprintf("%s?%s", BaseURI, param.Encode())

:)

Reply:
The only way I could see this breaking is if params were added to the base URI at some point, but that's highly unlikely so concurred - it feels like we're duplicating functionality a bit here but who cares it's definitely simpler, 12d787558234be15661bea9fad2cd78d8227722f